File: 3748-sort-matrix-by-diagonals/3748-sort-matrix-by-diagonals.go
package main

import (
	"sort"
)

func sortMatrix(grid [][]int) [][]int {
	n := len(grid)

	for x := 0; x < n; x++ {
		var r, c []int

		for i, j := x, 0; i < n && j < n; i, j = i+1, j+1 {
			r = append(r, grid[i][j])
		}
		for i, j := 0, x; i < n && j < n; i, j = i+1, j+1 {
			c = append(c, grid[i][j])
		}

		sort.Sort(sort.Reverse(sort.IntSlice(r)))
		sort.Ints(c)

		for i, j := 0, x; i < n && j < n; i, j = i+1, j+1 {
			grid[i][j] = c[i]
		}
		for i, j := x, 0; i < n && j < n; i, j = i+1, j+1 {
			grid[i][j] = r[j]
		}
	}

	return grid
}

File: 3748-sort-matrix-by-diagonals/3748-sort-matrix-by-diagonals.cpp
class Solution {
public:
    vector<vector<int>> sortMatrix(vector<vector<int>>& grid) {
        int n = grid.size();
        for(int x = 0; x < n; x++) {
            vector<int> r,c;
            for(int i = x, j = 0; i < n and j < n; i++,j++) {
                r.push_back(grid[i][j]);
            }
            for(int i = 0, j = x; i < n and j < n; i++,j++) {
                c.push_back(grid[i][j]);
            }
            
            sort(rbegin(r), rend(r));
            sort(begin(c), end(c));
            
            for(int i = 0, j = x; i < n and j < n; i++,j++) {
                grid[i][j] = c[i];
            }
            for(int i = x, j = 0; i < n and j < n; i++,j++) {
                grid[i][j] = r[j];
            }
            
        }
        return grid;
    }
};